File: phase0-data-structures/arrays-dynamic-array/Solution.cs
var arr = new DynamicArray<int>(2);
Console.WriteLine("Pushing 1..6 into a DynamicArray with initial capacity 2:");
for (int i = 1; i <= 6; i++) arr.Push(i);
Console.WriteLine($"Final contents: [{string.Join(", ", arr.ToArray())}]");

arr.InsertAt(2, 99);
Console.WriteLine($"After InsertAt(2, 99): [{string.Join(", ", arr.ToArray())}]");

arr.DeleteAtSwap(0);
Console.WriteLine($"After DeleteAtSwap(0): [{string.Join(", ", arr.ToArray())}]");

class DynamicArray<T>
{
    private T[] data;
    private int count;
    private int capacity;

    public DynamicArray(int initialCapacity = 4)
    {
        capacity = initialCapacity;
        data = new T[capacity];
        count = 0;
    }

    public int Size => count;

    public T Get(int index)
    {
        if (index < 0 || index >= count) throw new IndexOutOfRangeException($"Index {index} out of bounds");
        return data[index];
    }

    public void Push(T value)
    {
        if (count == capacity) Resize();
        data[count] = value;
        count++;
    }

    public void DeleteAtSwap(int index)
    {
        if (index < 0 || index >= count) throw new IndexOutOfRangeException($"Index {index} out of bounds");
        data[index] = data[count - 1];
        data[count - 1] = default!;
        count--;
    }

    public void InsertAt(int index, T value)
    {
        if (index < 0 || index > count) throw new IndexOutOfRangeException($"Index {index} out of bounds");
        if (count == capacity) Resize();
        for (int i = count; i > index; i--) data[i] = data[i - 1];
        data[index] = value;
        count++;
    }

    private void Resize()
    {
        capacity *= 2;
        var newData = new T[capacity];
        Array.Copy(data, newData, count);
        data = newData;
        Console.WriteLine($"  [resized to capacity {capacity}]");
    }

    public T[] ToArray()
    {
        var result = new T[count];
        Array.Copy(data, result, count);
        return result;
    }
}

File: phase0-data-structures/arrays-dynamic-array/solution.ts
class DynamicArray<T> {
    private data: (T | undefined)[];
    private count: number;
    private capacity: number;

    constructor(initialCapacity: number = 4) {
        this.capacity = initialCapacity;
        this.data = new Array(this.capacity);
        this.count = 0;
    }

    get size(): number  {
        return this.count;
    }

    get(index: number): T {
        if (index < 0 || index >= this.count) {
            throw new RangeError("Index out of bounds");
        }

        return this.data[index] as T;
    }

    push(value: T): void {
        if (this.count === this.capacity) {
            this.resize();
        }

        this.data[this.count] = value;
        this.count++;
    }

    deleteAtSwap(index: number): void {
        if (index < 0 || index >= this.count) {
            throw new RangeError("Index out of bounds");
        }
        this.data[index] = this.data[this.count-1];
        this.data[this.count - 1] = undefined;
        this.count--;
    }

    insertAt(index: number, value: T): void {
        if (index < 0 || index > this.count) {
            throw new RangeError("Index out of bounds");
        }

        if (this.count === this.capacity) {
            this.resize();
        }

        for (let i = this.count; i > index; i--) {
            this.data[i] = this.data[i - 1];
        }

        this.data[index] = value;
        this.count++;
    }

    private resize(): void {
        this.capacity *= 2;
        const newData = new Array(this.capacity);
        for(let i = 0; i < this.count; i++) {
            newData[i] = this.data[i];
        }
        this.data = newData;
        console.log(`Resized array to new capacity: ${this.capacity}`);
    }


    toArray(): T[] {
        return this.data.slice(0, this.count) as T[];
    }

}


const arr = new DynamicArray<number>(2);
console.log("Pushing 1..6 into a DynamicArray with initial capacity 2");
for (let i = 1; i <= 6; i++) arr.push(i);
console.log("Final content of the array:", arr.toArray());

arr.insertAt(2, 99);
console.log("After inserting 99 at index 2:", arr.toArray());

arr.deleteAtSwap(0);
console.log("After deleting element at index 0 (swap with last):", arr.toArray());